File: packages/sdk-typescript/src/endpoint.ts
/**
 * Transport-plurality endpoint descriptor + serial-sticky negotiation.
 *
 * Sprint 2 (wire v1.3.0-beta.1): a transfer carries a `reachable_at[]`
 * array of endpoints. The recipient tries them in the sender's declared
 * priority order (ADR-0012: sender-ranked, serial, sticky) and stops at
 * the first that answers. Stickiness is per-transfer: subsequent
 * transfers re-evaluate from the top.
 *
 * Unknown `kind` values are preserved on the wire but SKIPPED during
 * fallback — a forward-compatible peer on a newer wire can advertise a
 * transport this SDK can't speak, and we gracefully fall through
 * instead of exploding.
 */

// Keep these in sync with `aex_core::Endpoint::KIND_*` and the Python
// SDK's `aex_sdk.endpoint` constants.
export const KIND_CLOUDFLARE_QUICK = "cloudflare_quick";
export const KIND_CLOUDFLARE_NAMED = "cloudflare_named";
export const KIND_IROH = "iroh";
export const KIND_TAILSCALE_FUNNEL = "tailscale_funnel";
export const KIND_FRP = "frp";

export const KNOWN_KINDS: ReadonlySet<string> = new Set([
  KIND_CLOUDFLARE_QUICK,
  KIND_CLOUDFLARE_NAMED,
  KIND_IROH,
  KIND_TAILSCALE_FUNNEL,
  KIND_FRP,
]);

/**
 * Endpoint kinds this SDK can dial today. Iroh requires a QUIC client
 * wire-up that lands in a later PR; for now Iroh entries are skipped
 * during fallback with a `skippedReason` instead of dialed.
 */
export const HTTP_KINDS: ReadonlySet<string> = new Set([
  KIND_CLOUDFLARE_QUICK,
  KIND_CLOUDFLARE_NAMED,
  KIND_TAILSCALE_FUNNEL,
  KIND_FRP,
]);

export interface Endpoint {
  kind: string;
  url: string;
  priority: number;
  healthHintUnix?: number;
}

/** JSON shape as stored on the wire (snake_case). */
export interface EndpointJson {
  kind: string;
  url: string;
  priority?: number;
  health_hint_unix?: number | null;
}

export function endpointFromJson(obj: EndpointJson): Endpoint {
  return {
    kind: obj.kind,
    url: obj.url,
    priority: obj.priority ?? 0,
    healthHintUnix:
      obj.health_hint_unix === null || obj.health_hint_unix === undefined
        ? undefined
        : Number(obj.health_hint_unix),
  };
}

export function endpointToJson(ep: Endpoint): EndpointJson {
  const out: EndpointJson = {
    kind: ep.kind,
    url: ep.url,
    priority: ep.priority,
  };
  if (ep.healthHintUnix !== undefined) {
    out.health_hint_unix = ep.healthHintUnix;
  }
  return out;
}

export function isKnownKind(ep: Endpoint): boolean {
  return KNOWN_KINDS.has(ep.kind);
}

export function isHttpDialable(ep: Endpoint): boolean {
  return HTTP_KINDS.has(ep.kind);
}

/** Return a new array sorted by priority (lower first). `Array.sort`
 * is stable so ties keep insertion order. */
export function sortByPriority(endpoints: readonly Endpoint[]): Endpoint[] {
  return [...endpoints].sort((a, b) => a.priority - b.priority);
}

export interface FallbackAttempt {
  endpoint: Endpoint;
  error?: string;
  skippedReason?: string;
  ok: boolean;
}

export interface FallbackResult<T> {
  value?: T;
  chosen?: Endpoint;
  attempts: FallbackAttempt[];
}

export function succeeded<T>(result: FallbackResult<T>): result is FallbackResult<T> & {
  value: T;
  chosen: Endpoint;
} {
  return result.chosen !== undefined;
}

/**
 * Walk `endpoints` in sender-declared priority order and invoke
 * `attempt(endpoint)` on each until one returns without throwing.
 *
 * - Unknown / non-dialable kinds are skipped with a `skippedReason`;
 *   they never count as failures for the at-least-one-succeeded
 *   invariant.
 * - The first rejection from a known-dialable endpoint is recorded
 *   but does NOT abort fallback — we keep walking.
 * - Once one attempt resolves, the loop stops (sticky per-transfer per
 *   ADR-0012).
 */
export async function tryEndpoints<T>(
  endpoints: readonly Endpoint[],
  attempt: (ep: Endpoint) => Promise<T>,
  options?: { onSkip?: (ep: Endpoint, reason: string) => void },
): Promise<FallbackResult<T>> {
  const attempts: FallbackAttempt[] = [];
  const ordered = sortByPriority(endpoints);
  for (const ep of ordered) {
    if (!isKnownKind(ep)) {
      const reason = `unknown kind: '${ep.kind}'`;
      attempts.push({ endpoint: ep, skippedReason: reason, ok: false });
      options?.onSkip?.(ep, reason);
      continue;
    }
    if (!isHttpDialable(ep)) {
      const reason = `${ep.kind} is not HTTP-dialable from this SDK yet`;
      attempts.push({ endpoint: ep, skippedReason: reason, ok: false });
      options?.onSkip?.(ep, reason);
      continue;
    }
    try {
      const value = await attempt(ep);
      attempts.push({ endpoint: ep, ok: true });
      return { value, chosen: ep, attempts };
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      attempts.push({ endpoint: ep, error: message, ok: false });
    }
  }
  return { attempts };
}

File: packages/sdk-python/src/aex_sdk/endpoint.py
"""Transport-plurality endpoint descriptor + serial-sticky negotiation.

Sprint 2 (wire v1.3.0-beta.1): a transfer carries a ``reachable_at[]``
array of endpoints. The recipient tries them in the sender's declared
priority order (ADR-0012: sender-ranked, serial, sticky) and stops at
the first that answers. Stickiness is per-transfer: subsequent transfers
re-evaluate from the top.

Unknown ``kind`` values are preserved on the wire but SKIPPED during
fallback — a forward-compatible peer on a newer wire can advertise a
transport this SDK can't speak, and we gracefully fall through instead
of exploding.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Iterable, Optional

# Known transport kinds. Keep in sync with the Rust
# `aex_core::Endpoint::KIND_*` constants.
KIND_CLOUDFLARE_QUICK = "cloudflare_quick"
KIND_CLOUDFLARE_NAMED = "cloudflare_named"
KIND_IROH = "iroh"
KIND_TAILSCALE_FUNNEL = "tailscale_funnel"
KIND_FRP = "frp"

KNOWN_KINDS: frozenset[str] = frozenset(
    {
        KIND_CLOUDFLARE_QUICK,
        KIND_CLOUDFLARE_NAMED,
        KIND_IROH,
        KIND_TAILSCALE_FUNNEL,
        KIND_FRP,
    }
)

# Endpoints this SDK can actually dial today. Iroh requires an iroh
# client wire-up that lands in a later PR; until then Iroh endpoints are
# skipped with a debug log rather than dialed.
HTTP_KINDS: frozenset[str] = frozenset(
    {
        KIND_CLOUDFLARE_QUICK,
        KIND_CLOUDFLARE_NAMED,
        KIND_TAILSCALE_FUNNEL,
        KIND_FRP,
    }
)


@dataclass(frozen=True)
class Endpoint:
    """A single way to reach a sender's data plane."""

    kind: str
    url: str
    priority: int = 0
    health_hint_unix: Optional[int] = None

    @classmethod
    def from_json(cls, obj: dict[str, Any]) -> "Endpoint":
        return cls(
            kind=str(obj["kind"]),
            url=str(obj["url"]),
            priority=int(obj.get("priority", 0)),
            health_hint_unix=(
                int(obj["health_hint_unix"]) if obj.get("health_hint_unix") is not None else None
            ),
        )

    def to_json(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "kind": self.kind,
            "url": self.url,
            "priority": self.priority,
        }
        if self.health_hint_unix is not None:
            out["health_hint_unix"] = self.health_hint_unix
        return out

    @property
    def is_known_kind(self) -> bool:
        return self.kind in KNOWN_KINDS

    @property
    def is_http_dialable(self) -> bool:
        return self.kind in HTTP_KINDS


def sort_by_priority(endpoints: Iterable[Endpoint]) -> list[Endpoint]:
    """Return endpoints ordered by sender priority (lower first), ties
    broken by original position (Python's sort is stable)."""
    return sorted(endpoints, key=lambda e: e.priority)


@dataclass
class FallbackAttempt:
    """One attempt in [`try_endpoints`] — useful for surfacing diagnostics
    to the caller when all transports fail."""

    endpoint: Endpoint
    error: Optional[str] = None
    skipped_reason: Optional[str] = None
    ok: bool = False


@dataclass
class FallbackResult:
    """Outcome of [`try_endpoints`]. On success ``value`` is whatever the
    per-endpoint callable returned for the first working endpoint."""

    value: Any = None
    chosen: Optional[Endpoint] = None
    attempts: list[FallbackAttempt] = field(default_factory=list)

    @property
    def succeeded(self) -> bool:
        return self.chosen is not None


def try_endpoints(
    endpoints: Iterable[Endpoint],
    attempt: Callable[[Endpoint], Any],
    *,
    on_skip: Optional[Callable[[Endpoint, str], None]] = None,
) -> FallbackResult:
    """Walk ``endpoints`` in sender-declared priority order; invoke
    ``attempt(endpoint)`` on each until one returns without raising.

    - Unknown / non-dialable kinds are skipped with a recorded reason;
      they never count as failures for the purposes of the
      at-least-one-succeeded invariant.
    - The first exception from a known-dialable endpoint is recorded but
      does NOT abort fallback — we keep walking.
    - Once one attempt succeeds, the loop stops (sticky per-transfer per
      ADR-0012).

    Returns a [`FallbackResult`] populated with the chosen endpoint, the
    attempt's return value, and per-endpoint diagnostics.
    """
    result = FallbackResult()
    ordered = sort_by_priority(endpoints)
    for ep in ordered:
        if not ep.is_known_kind:
            reason = f"unknown kind: {ep.kind!r}"
            result.attempts.append(FallbackAttempt(endpoint=ep, skipped_reason=reason))
            if on_skip is not None:
                on_skip(ep, reason)
            continue
        if not ep.is_http_dialable:
            reason = f"{ep.kind} is not HTTP-dialable from this SDK yet"
            result.attempts.append(FallbackAttempt(endpoint=ep, skipped_reason=reason))
            if on_skip is not None:
                on_skip(ep, reason)
            continue
        try:
            value = attempt(ep)
        except Exception as exc:  # pragma: no cover - the per-endpoint
            # error path is covered in test_endpoint.py with a mock that
            # raises; the `except Exception` is deliberately broad so a
            # failing endpoint never crashes fallback.
            result.attempts.append(FallbackAttempt(endpoint=ep, error=str(exc)))
            continue
        result.attempts.append(FallbackAttempt(endpoint=ep, ok=True))
        result.value = value
        result.chosen = ep
        return result
    return result
